src/helpers.rs

```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                                res: hir::def::Res::PrimTy(hir::PrimTy::Int(IntTy::I64)), ..},..)
+                            ), ..
+                    }) = (k.to_i64(), arg_type) {
+                            libffi_args.push((Box::new(v), CArg::Int64(v)));
+                    }
+                    // the uints
+                    else if let (Ok(v), &hir::Ty{
+                        hir_id:_, kind: hir::TyKind::Path(
+                            hir::QPath::Resolved(_, hir::Path { 
+                                span: _, 
+                                res: hir::def::Res::PrimTy(hir::PrimTy::Uint(UintTy::U8)), ..},..)
+                            ), ..
+                    }) = (k.to_u8(), arg_type) {
+                            libffi_args.push((Box::new(v), CArg::UInt8(v)));
+                    } else if let (Ok(v), &hir::Ty{
+                        hir_id:_, kind: hir::TyKind::Path(
+                            hir::QPath::Resolved(_, hir::Path { 
+                                span: _, 
+                                res: hir::def::Res::PrimTy(hir::PrimTy::Uint(UintTy::U16)), ..},..)
+                            ), ..
+                    }) = (k.to_u16(), arg_type) {
+                            libffi_args.push((Box::new(v), CArg::UInt16(v)));
+                    } else if let (Ok(v), &hir::Ty{
+                        hir_id:_, kind: hir::TyKind::Path(
+                            hir::QPath::Resolved(_, hir::Path { 
+                                span: _, 
+                                res: hir::def::Res::PrimTy(hir::PrimTy::Uint(UintTy::U32)), ..},..)
+                            ), ..
+                    }) = (k.to_u32(), arg_type) {
+                            libffi_args.push((Box::new(v), CArg::UInt32(v)));
+                    } else if let (Ok(v), &hir::Ty{
+                        hir_id:_, kind: hir::TyKind::Path(
+                            hir::QPath::Resolved(_, hir::Path { 
+                                span: _, 
+                                res: hir::def::Res::PrimTy(hir::PrimTy::Uint(UintTy::U64)), ..},..)
+                            ), ..
+                    }) = (k.to_u64(), arg_type) {
+                            libffi_args.push((Box::new(v), CArg::UInt64(v)));
+                    } else {
+                        libffi_args.push((Box::new(()), CArg::INVALID));
+                        throw_unsup_format!("Unsupported scalar argument type to external C function: {:?}", cur_arg);
+                    }
+                },
+                _ => {
+                    libffi_args.push((Box::new(()), CArg::INVALID));
+                    throw_unsup_format!("Unsupported argument type to external C function: {:?}", cur_arg);
+                }
+            }
+        }
+
+        // convert them to libffi Arg type
+        let libffi_args = libffi_args.iter()
+            .map(|cur_arg| {
+                match cur_arg {
+                    (boxed, CArg::Int8(_)) => {
```

Comment:
Probably can be factored out into a method on `CArg` that projects into a libffi arg?

Reply:
yes, good idea!